### ISSUE TITLE
preloading a CSS image

### ISSUE THREAD

**Issue by Zhang-Junzhi** [w3c/csswg-drafts#2994]
For now, we seem to have to do something _hacky_ if we want to preload a image for use as a CSS image later. Despite some hacky workarounds requiring script or document code invovled, there are some pure CSS wordaround(like the following example):

```
body::before
{
    content: url(foo-bg.png) url(bar-bg.png) url(other-bg1.png) url(other-bg2.png);
    position: absolute;
    height: 0px;
    width: 0px;
    overflow: hidden;
    z-index: -1;
}

#foo:hover
{
    background: url(foo-bg.png);
    ...
}

...
```

It seems preloading a CSS image is "prefectly" resolved via pure CSS, however, these pure CSS methods look very counter-intuitive, lengthy and maintenance-invovling, and even further, they aren't just _theoretically garanteed_ to be correct(though they work in most of broswers in the real world).

I'd really like to see a native CSS feature that explictly provides ability to preload a image.

Any takes on this issue? I'm looking forward to hearing from you.

**Comment by upsuper** [comment]
Maybe a [`<url-modifier>` ](https://drafts.csswg.org/css-values-3/#url-modifiers) can be defined to do this.

**Comment by tabatkins** [comment]
Yeah, that's probably the most reasonable way to target this.

**Comment by Zhang-Junzhi** [comment]
Thanks for your opinions. I love the idea. Maybe speical cases like url(attr(...)) needs to be treated specially.

**Comment by inoas** [comment]
Could there be made a distinction between preload (high priority, before regular requests), load (with regular requests) and prefetch (low priority, after regular requests)?

**Comment by Zhang-Junzhi** [comment]
Thanks for the reply. @inoas

To further break down image loading priorities, images with `load` priority can be devided into different sub-priorities.

For example,

```
<style>
body
{
	background: url(page-bg.png);
}

#page-loading-progress::before
{
	background: url(spinning-circle.png) ... no-repeat;
}
</style>
<body>
	<progress id="page-loading-progress">...</progress>
	...
</body>
```

Here it's reasonable to prioritise the load for the spinning-circle loading indicator as opposed to the large one that's used in `body`(though practically browsers tend to load both of the two images in parallel).

**Comment by Malvoz** [comment]
Priorities may be delegated using [Priority Hints](https://github.com/WICG/priority-hints).

**Comment by AmeliaBR** [comment]
Related: #1603 on crossorigin modifier and #2095 on async modifier. We should probably address them all at the same time to get consistent syntax and implementation guidelines.
